### PR TITLE
Content management ui forms: projects + environments

### PR DESCRIPTION
**Pull request by lneves12**
## What does this PR change?

Content management: projects + environments - UI Form + CRUD web api 

Permission management is still missing: https://github.com/SUSE/spacewalk/issues/7144

There is still some code cleanup and better error handling that can be done, but for an MVP and to merge into content-management it's in a good state already

**UI - Create**
cm-create-flow

**UI - Delete**
cm-delete-flow

## Changelogs

Copy the following sentence as a new comment if you don't need changelog entries:

> gitarro no changelog needed !!!

If the test `changelog_test`already run, then add another new comment with the following text:

> gitarro rerun changelog_test !!!

### CODE REVIEW

**Comment by lneves12** [comment]
gitarro no changelog needed !!!

**Comment by lneves12** [comment]
Besides some javadoc java linting errors, this PR is already in a good state to be reviewed

File: branding/java/code/src/com/redhat/rhn/branding/strings/StringResource_en_US.xml [hunk 1]
```diff
@@ -2081,6 +2081,14 @@ tree will not be usable for autoinstalling.
         <source>Provisioning is currently not supported for systems managed with push via SSH tunnel.</source>
       </trans-unit>
 
+      <!-- Content Management -->
```
[the remark this answers is not in the document]
Comment (lneves12) [reply branding/java/code/src/com/redhat/rhn/branding/strings/StringResource_en_US.xml]: that's a good question. cc @ncounter 

Anyway, if everything is migrated to react, I would say that this file will eventually disappear and everything will be managed by the function "t()" in react

File: web/html/src/components/input/Label.js [hunk 2]
```diff
@@ -17,7 +17,6 @@ function Label(props: Props) {
     >
       { props.name }
       { props.required ? <span className="required-form-field"> *</span> : undefined }
-            :
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/components/input/Label.js]: I will revert it, I see all the old code is using that ":"

IMO It looks quite ugly and useless, but even if we remove them it should in another PR

Screenshot from 2019-03-12 00-24-26
 cc @ncounter

File: web/html/src/manager/content-management/project/project.js [hunk 3]
```diff
@@ -1,22 +1,122 @@
...
+    if(props.wasFreshlyCreatedMessage) {
+      showSuccessToastr(props.wasFreshlyCreatedMessage)
+    }
+  }, [])
+
+  const editedStates = ["0","2","3"];
+  const statesDesc = {
+    "0": "added",
+    "1": "built",
+    "2": "edited",
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/manager/content-management/project/project.js]: This will be removed on the sources UI PR, all the logic will be implemented on the server side. If you prefer I can remove all the sources logic from this PR. 

Anyway, I think we will still need the state Edited for filters in case we reuse the same enum.

We also talked about the fact to give some visual feedback that a project needs to be rebuilt when an "mgr-sync" is run and a source gets new packages, we could use that edited (yellow) state to represent that.

File: web/html/src/manager/content-management/project/project.renderer.js [hunk 4]
```diff
@@ -5,9 +5,17 @@ import Project from './project';
...
+  let projectJson = {};
+  try{
+    projectJson = JSON.parse(project);
+  }  catch(error) {}
+
   ReactDOM.render(
-    <Project />,
+    <Project
+      project={projectJson}
+      { ...( wasFreshlyCreatedMessage && { wasFreshlyCreatedMessage } ) }
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/manager/content-management/project/project.renderer.js]: ahah kind of :P
 Anyway I will simplify it with https://github.com/lneves12/uyuni/blob/56c202c179feecb8229f7ac72a06eb5506f4a029/web/html/src/manager/content-management/list-projects/list-projects.renderer.js#L18 and change the name to "flashMessage" too.

That's basically passing optional parameters through destructuring to <Project />.
If wasFreshlyCreatedMessage is false it will not pass any prop and if it's defined it will pass wasFreshlyCreatedMessage={wasFreshlyCreatedMessage}.

File: web/html/src/manager/errors/default-error-500.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+
+const DefaultError500 = () =>
+  <body>
+    <h1>
+      Internal Server Error
+    </h1>
+    <p>
+      Please reload the page.
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/manager/errors/default-error-500.js]: To avoid confusions I will change this component name to DefaultJsError

This page is a bit different than the 500.jsp one, it doesn't refer to a server internal error.

Basically, prior React 16 the default behavior of React whenever there is a js error while executing the render is to show a blank page. This behavior used to be different on the react versions < 16, it would keep showing stuff on the UI even if the state is inconsistent.
 
In the new React 16 version, a new feature "errorBoundaries" was introduced in order to have a more reliable error handling. With that feature, we can create some sort of try catchs throughout our components tree and react to error in different ways depending where the error occurs, for instance if an error happens in a deep nested panel we can catch it there and try to recover smoothly instead of disabling the whole page.

This component is the default page if any of those errors isn't catched. This is being defined here https://github.com/lneves12/uyuni/blob/2ca9e651ad68f69326ef66cd24f504c1c3875684/web/html/src/components/general/with-page-wrapper.js#L15

File: java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/EnvironmentHandler.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    public static HashMap<String,String> validateEnvironmentRequest(EnvironmentRequest envRequest) {
+        HashMap<String,String> requestErrors = new HashMap<>();
+
+        if(StringUtils.isEmpty(envRequest.getLabel())) {
+            requestErrors.put("label", "Label is Required");
+        }
+
+        if(StringUtils.isEmpty(envRequest.getName())) {
+            requestErrors.put("name", "Label is Required");
```
[the remark this answers is not in the document]
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/EnvironmentHandler.java]: well caught :)

File: web/html/src/manager/content-management/project/project.js [hunk 7]
```diff
@@ -62,22 +62,22 @@ const Project = (props: Props) => {
...
-                            .catch((error) => {
-                              showErrorToastr(error);
-                            })
-                        }
-          />
         </div>
       }
     >
+      <DeleteDialog id="delete-project-modal"
+          title={t("Delete Store")}
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/manager/content-management/project/project.js]: well caught :) Probably some copy paste left overs

File: java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/ProjectApiController.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        } catch (ContentManagementException error) {
+            return json(GSON, res, HttpStatus.SC_BAD_REQUEST, ResultJson.error(error.getMessage()));
+        }
+
+        FlashScopeHelper.flash(
+                req,
+                String.format("Project %s created successfully.", projectPropertiesRequest.getLabel())
+        );
+
+        List<ContentEnvironment> contentEnvironments = ContentManager.listProjectEnvironments(createdProject.getLabel(), user);
```
[the remark this answers is not in the document]
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/contentmanagement/handlers/ProjectApiController.java]: right :) 

Since this is also being reuse in so many places I think I will create something like:

`return fullProjectJsonResponse(projectLabel, user)` and centralize this code there.

What do you think?

File: java/code/src/com/suse/manager/webui/controllers/contentmanagement/ContentManagementViewsController.java [hunk 9]
```diff
@@ -33,57 +47,54 @@
...
-     * @return the model and view
-     */
-    public static ModelAndView projectView(Request req, Response res) {
+    public static ModelAndView listProjectsView(Request req, Response res, User user) {
         Map<String, Object> data = new HashMap<>();
-        return new ModelAndView(data, "controllers/contentmanagement/templates/project.jade");
+
+        List<ContentProject> projects = ContentManager.listProjects(user);
+        Map<ContentProject, List<ContentEnvironment>> environmentsByProject = new HashMap<>();
+        projects.forEach(project -> environmentsByProject.put(project, ContentManager.listProjectEnvironments(project.getLabel(), user)));
```
[the remark this answers is not in the document]
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/contentmanagement/ContentManagementViewsController.java]: uh thanks for the good tip :)

Still learning all the "new" java stream ticks

File: java/code/src/com/suse/manager/webui/controllers/contentmanagement/mappers/ResponseMappers.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        project.setProperties(mapProjectPropertiesFromDB(projectDB));
+        project.setSources(sources);
+        project.setEnvironments(mapEnvironmentsFromDB(envsDB));
+
+        return project;
+    }
+
+    public static List<ProjectResumeResponse> mapProjectListingFromDB(Map<ContentProject, List<ContentEnvironment>> envsByProjDB) {
+        List<ProjectResumeResponse> projectsResume = new ArrayList<ProjectResumeResponse>();
+        envsByProjDB.forEach((project, environments) -> {
```
[the remark this answers is not in the document]
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/contentmanagement/mappers/ResponseMappers.java]: :+1:

File: java/code/src/com/suse/manager/webui/controllers/contentmanagement/ContentManagementApiController.java [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                removeEnvironmentRequest.getProjectLabel(),
+                user
+        );
+
+        Optional<ContentProject> updatedProject = ContentManager.lookupProject(
+                removeEnvironmentRequest.getProjectLabel(),
+                user
+        );
+
+        if(updatedProject.isPresent()) {
```
[the remark this answers is not in the document]
Comment (lneves12) [reply java/code/src/com/suse/manager/webui/controllers/contentmanagement/ContentManagementApiController.java]: Thanks for the  tip, it indeed looks like a better use of the Optional operator :)

File: web/html/src/utils/network.js [hunk 12]
```diff
@@ -52,7 +52,7 @@ function errorMessageByStatus(status) {
     } else if (status >= 500) {
         return [t("Server error, please check log files.")];
     } else {
-        return [];
+        return [t("Server error, please check log files.")];
```
[the remark this answers is not in the document]
Comment (lneves12) [reply web/html/src/utils/network.js]: I changed this to: 

```javascript
    if (status === 401) {
        return [t("Session expired, please reload the page.")];
    } else if (status === 403) {
        return [t("Authorization error, please reload the page or try to logout/login again.")];
    } else if (status >= 500) {
        return [t("Server error, please check log files.")];
    } else if (status === 400) {
      return [t("Server error, please check log files.")];
    } else {
      return [t("")];
    }
```

File: web/html/src/utils/network.js [hunk 13]
```diff
@@ -51,8 +51,10 @@ function errorMessageByStatus(status) {
         return [t("Authorization error, please reload the page or try to logout/login again.")];
     } else if (status >= 500) {
         return [t("Server error, please check log files.")];
+    } else if (status === 400) {
```
Comment (hustodemon) [inline web/html/src/utils/network.js]: Sorry for being picky here :smile: I think the errors between 400 and 499 are **client** errors, not server ones. Moreover, I think that differentiating message for this one would be good  (e.g. report "Client error" for 400 and "server error for >= 500").

Comment (lneves12) [reply web/html/src/utils/network.js]: well yeah, but for the customer everything is a server error, IMO that is some kind of internal definition, for them, I would say it's hard to distinguish what is the difference between a client/server error. Anyway, I removed that and kept the old behavior, most of the cases we want to have custom error handling for 400 and show something like "Check all the input fields errors bellow"

File: web/html/src/components/input/Label.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ncounter) [reply web/html/src/components/input/Label.js]: Ah, I see. Yeah it is quite ugly in the most of the cases, but sometimes we don't have the `input` field right after, but a simple following text and the `column` is kind of needed to split the two parts. It could be definitely designed better, but it is too much spread around to softly remove it and *see what happens after*. Let's keep it and we will tackle this later on a UI/design round. :wink:

File: branding/java/code/src/com/redhat/rhn/branding/strings/StringResource_en_US.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncounter) [reply branding/java/code/src/com/redhat/rhn/branding/strings/StringResource_en_US.xml]: Ah, more than a good question. Nothing decided on top of that. I would continue adding stuff into `branding` in a way we can still distinguish what comes from the *old redhat spacewalk upstream* project and what comes from us, *suma internal developed only*. Therefore I expect only *branding* will be touched from now on both on SUSE-Manager and Uyuni repos. Later on when we will squash everything, we will restructure this side as well from my POV.

File: java/code/src/com/suse/manager/webui/menu/MenuTree.java [hunk 14]
```diff
@@ -231,8 +231,10 @@
 
             // Content Management
             nodes.add(new MenuItem("contentmanagement.nav.title").withIcon("spacewalk-icon-software-channels")
-                .addChild(new MenuItem("contentmanagement.nav.list")
-                        .withPrimaryUrl("/rhn/manager/contentmanagement/projects")));
+                    .withDir("/rhn/manager/contentmanagement")
```
[the remark this answers is not in the document]
Comment (ncounter) [reply java/code/src/com/suse/manager/webui/menu/MenuTree.java]: tl;dr : `withDir` is a fallback **partial** match of the URL that looks for an active menu item to select. I saw many times not having a fallback `withDir` reflected into *no-menu-item-active* result.

---

The `withDir` means *everything under this `path` that has no better specific detected URL MenuItem* it will be reflected in the menu as belonging to this `MenuItem` even if it does not match perfectly. It's like a wrapper to say *if you do not find anything that matches with the current URL, mark this MenuItem as selected as a fallback*. The same is happening also at [L237](https://github.com/uyuni-project/uyuni/pull/430/files/#diff-99972ff1af9cd70a9cfcb95a01374f18R237).

For instance:
- request is `/rhn/manager/contentmanagement/project/my-test-project`
- no **perfect** URL MenuItem is found
- the [`getBestActiveDir`](https://github.com/uyuni-project/uyuni/blob/dae038788ce3a63c92ae3d9d45cec11cebcecb7b/java/code/src/com/suse/manager/webui/menu/MenuTree.java#L502) method will look for the *best partial matching* MenuItem

In the `.withDir("/rhn/manager/contentmanagement/project")` case it will mark as `active` the `new MenuItem("contentmanagement.nav.projects")` element (second level) of the menu in any single Project page. In the `.withDir("/rhn/manager/contentmanagement/")` it will mark the `new MenuItem("contentmanagement.nav.title")` element (first level) of the menu. The latter I don't expect to happen but in case we will add any additional URL/Page without a specific MenuItem, we already have a fallback.

**Comment by lneves12** [comment]
@ncounter I am not sure about the effort of migrating all the other pages to use the new notification system, it might be a lot of work specially on older pages, but we can try to evaluate that...
In the beginning, I tried to reuse the current notifications process, but it's not an easy task as well. You have a lot of operations happening inside modals with most likely a big scroll on the page, showing a static notification on top of the page can be tricky too.

**Comment by lneves12** [comment]
I adapted the new notifications style to be the same as the rest of the page. https://github.com/uyuni-project/uyuni/pull/430/commits/db1096b675ff310cca116ddeea980529833c0fd8

Example:

notifications_style

@ncounter @mcalmer
